### PR TITLE
Upgraded the module to meet gpt4all-ui backend requirements

### PR DESCRIPTION
**Pull request by ParisNeo**
Added the possibility to stop the generation process any time.
Added a function to count the number of tokens that a prompt has. This is needed by GPT4All-ui to optimize the generation process.

Thanks alot

### CODE REVIEW

**Comment by abdeladim-s** [comment]
Thanks @ParisNeo

File: pyllamacpp/pyllamacpp/model.py [hunk 1]
```diff
@@ -75,21 +75,35 @@ def _set_params(params, kwargs: dict) -> None:
...
         # save res
         self.res += text
+        return True
 
     def _call_grab_text_callback(self) -> str:
         if Model._grab_text_callback is not None:
             return Model._grab_text_callback()
         return None
 
+    def num_tokens(self, prompt:str):
```
Comment (abdeladim-s) [inline pyllamacpp/pyllamacpp/model.py:96]: @ParisNeo Where this function is used ?

File: pyllamacpp/src/main.cpp [hunk 2]
```diff
@@ -447,7 +455,11 @@ int llama_generate(struct llama_context_wrapper * ctx_w, gpt_params params, py::
         if (!input_noecho) {
             for (auto id : embd) {
 //                printf("%s", llama_token_to_str(ctx, id));
-                new_text_callback(llama_token_to_str(ctx, id));
+                // If the host wants to stop generation, we should stop
+                if(!new_text_callback(llama_token_to_str(ctx, id))){
```
Comment (abdeladim-s) [inline pyllamacpp/src/main.cpp]: You are calling `new_text_callback` two times. 
Is this the right behavior !?

**Comment by abdeladim-s** [comment]
@ParisNeo as I told you, I think you can do the same with version 1.0.7, have you tried it ?

File: pyllamacpp/src/main.cpp (hunk 2 above)
Comment (ParisNeo) [reply pyllamacpp/src/main.cpp]: nope. 

Also this is not my code. Here is my code:
                new_text_callback(llama_token_to_str(ctx, id));
                if(!ctx_w->continue_gen){
                    llama_print_timings(ctx);
                    return 0;
                }

Comment (ParisNeo) [reply pyllamacpp/src/main.cpp]: Can you take a look at the Files changed tab of my pull request?
You seem to be looking at an old commit. I didn't rebase my commits before pushing so if you're not looking at the last one, you may find funny things I was testing.

And I didn't understand what do you mean by doing this in 1.0.7?

**Pull request closed without merging**